### PR TITLE
Upgraded factory

### PR DESCRIPTION
**Pull request by Remscar**

### CODE REVIEW

**Review by MattCharles — APPROVED** [review]

File: contracts/zStakePoolFactory.sol [hunk 1]
```diff
@@ -237,7 +237,7 @@ contract zStakePoolFactory is OwnableUpgradeable, PausableUpgradeable {
    *
    * @param perBlock Amount of wild given per block
    */
-  function changeRewardTokensPerBlock(uint256 perBlock) external {
+  function changeRewardTokensPerBlock(uint256 perBlock) external onlyOwner {
```
Comment (MattCharles) [inline contracts/zStakePoolFactory.sol:240]: Yikes

**Comment by MattCharles** [comment]
Would `changePoolWeight` need to change too? If a user has a deposit in a given pool, does that method allow them to change the weight of the pool for all users?

File: contracts/zStakePoolFactory.sol (hunk 1 above)
Comment (JamesEarle) [reply contracts/zStakePoolFactory.sol:240]: image

**Review by JamesEarle — APPROVED** [review]

**Comment by JamesEarle** [comment]
> Would `changePoolWeight` need to change too? If a user has a deposit in a given pool, does that method allow them to change the weight of the pool for all users?

If a user was able to change the weight of a pool (the ratio of that pools payout compared to all pools) then they could definitely increase their own returns, as well as everyone else in the pool. 

The require in that function covers this though
```
require(msg.sender == owner() || poolExists[msg.sender]);
```

**Comment by beardhero** [comment]
> > Would `changePoolWeight` need to change too? If a user has a deposit in a given pool, does that method allow them to change the weight of the pool for all users?
> 
> If a user was able to change the weight of a pool (the ratio of that pools payout compared to all pools) then they could definitely increase their own returns, as well as everyone else in the pool.
> 
> The require in that function covers this though
> 
> ```
> require(msg.sender == owner() || poolExists[msg.sender]);
> ```

This function is vulnerable - anybody with a pool can set the pool weight for any other pool.
It checks poolExists[msg.sender] and then sets the weight of poolAddr that was passed in, which could be any address.
and setWeight() doesn't protect from it.

This function should not take poolAddr as an argument - it seems its sufficient to use the msg.sender as the pool address.

**Comment by JamesEarle** [comment]
> > > Would `changePoolWeight` need to change too? If a user has a deposit in a given pool, does that method allow them to change the weight of the pool for all users?
> > 
> > 
> > If a user was able to change the weight of a pool (the ratio of that pools payout compared to all pools) then they could definitely increase their own returns, as well as everyone else in the pool.
> > The require in that function covers this though
> > ```
> > require(msg.sender == owner() || poolExists[msg.sender]);
> > ```
> 
> This function is vulnerable - anybody with a pool can set the pool weight for any other pool. It checks poolExists[msg.sender] and then sets the weight of poolAddr that was passed in, which could be any address. and setWeight() doesn't protect from it.
> 
> This function should not take poolAddr as an argument - it seems its sufficient to use the msg.sender as the pool address.

`poolExists` is a mapping that can only be added to by calling  `registerPool` which has `onlyOwner` which you can see earlier in this code

**Comment by beardhero** [comment]
> > > > Would `changePoolWeight` need to change too? If a user has a deposit in a given pool, does that method allow them to change the weight of the pool for all users?
> > > 
> > > 
> > > If a user was able to change the weight of a pool (the ratio of that pools payout compared to all pools) then they could definitely increase their own returns, as well as everyone else in the pool.
> > > The require in that function covers this though
> > > ```
> > > require(msg.sender == owner() || poolExists[msg.sender]);
> > > ```
> > 
> > 
> > This function is vulnerable - anybody with a pool can set the pool weight for any other pool. It checks poolExists[msg.sender] and then sets the weight of poolAddr that was passed in, which could be any address. and setWeight() doesn't protect from it.
> > This function should not take poolAddr as an argument - it seems its sufficient to use the msg.sender as the pool address.
> 
> `poolExists` is a mapping that can only be added to by calling `registerPool` which has `onlyOwner` which you can see earlier in this code

Yes but once a pool exists, it can change the weight of any other pool by passing in a poolAddr other than itself.

**Pull request merged**